### ISSUE TITLE
Canonical safe place to add special metadata to the plugin `build` object?

### ISSUE THREAD

**Issue by jameswomack** [evanw/esbuild#1065]
My team's building some tooling around ESBuild & we've come across the need to share information across all plugins within a set of related builds. Adding to the `build` argument to `setup` can work well for this, but we'd like to avoid extending `build` in a way that isn't supported. Similar to how the `detail` or `data` convention is used on `CustomEvent` instances, or `context` is used in Webpack plugins, we'd like to do something like the following:

```js
// in plugins/env.js
const envPlugin = {
  name: 'env',

  // `metadata` in this case would be sort of "protected" by esbuild
  // as a place to put metadata specific to tools/systems that build
  // on top of esbuild's plugin system—>especially cross-module metadata
  // that `pluginData` is not suited to
  setup({onResolve, onLoad, metadata: {somePluginOrchestratorNamespace}}) {
    onResolve({ filter: /^env$/ }, args => {
        // The following is contrived, but illustrates maintaining state that sits about the plugin itself
        somePluginOrchestratorNamespace.resolvedEnvPaths.push(args.path);
        return {
            path: args.path,
            namespace: 'env-ns',
        };
    });

    onLoad({ filter: /.*/, namespace: 'env-ns' }, () => ({
      contents: JSON.stringify(process.env),
      loader: 'json',
    }))
  },
}

// in plugin-orchestrator.js ->
// the metadata field is "enhanced" w/ a field shared across all plugin executions
const plugins = rawPlugins.map(rawPlugin => {
  const plugin = {...rawPlugin};
  plugin.setup = rawBuild => {
    const build = {...rawBuild, metadata: {
         ...metadata, 
        somePluginOrchestratorNamespace: {resolvedEnvPaths:[]}
      }
    };
    return rawPlugin.setup(build);
  };
  return plugin;
});
```

**Comment by zaydek** [comment]
@jameswomack What was the issue with `pluginData`?

**Comment by evanw** [comment]
> What was the issue with `pluginData`?

I think the idea is that they want something other than a per-file store.

> Adding to the `build` argument to `setup` can work well for this

Can it? I wouldn't expect this to work. Each `setup` function is called with a temporary single-use object, so this wouldn't allow multiple plugins to communicate:

https://github.com/evanw/esbuild/blob/4b28b9049f0ee04c1db77e733874f84957fe7e0a/lib/common.ts#L644

It's not totally clear to me what problem you're trying to solve. Calling it an "orchestrator namespace" doesn't tell me anything about how it's used. Can you say more about what you're trying to do? For example:

Do you just need some data inside your plugin to last for the lifetime of a `build` call? If so, creating a local variable inside `setup` is sufficient. The callbacks you register can just reference that local variable.

Do you need a common set of variables between multiple plugins? If so, you could potentially store those variables in a common module and import that module into all of your plugins.

Do you need to loosely coordinate between multiple independently-published plugins? If so, perhaps storing something on `globalThis` could be sufficient, maybe with `Symbol.for('esbuild')`? I assume that is how people generally solve these kinds of problems since that's what you do in the browser with `window`.

**Comment by jameswomack** [comment]
> Can it? I wouldn't expect this to work. Each setup function is called with a temporary single-use object, so this wouldn't allow multiple plugins to communicate:

Yes, it can & does in limited testing—because of an central plugin management module we've built to retrieve wrapped plugins in a way that gets shared across multiple `esbuild` builds

> Do you just need some data inside your plugin to last for the lifetime of a build call?

No. If that were the case, your suggestion sounds great (and is appreciated). We have recursive build calls to control file naming beyond what `esbuild` allows  & to do things like dependency de-duplication (among other things)

> It's not totally clear to me what problem you're trying to solve.

As you said, we need "something other than a per-file store"... & it needs to work in *all* plugins & across multiple builds. An example of why is to ingest translation information from multiple files in one plugin & then be able to output the aggregate of that plugin's data w/ another plugin—& again, for this to work across a set of `esbuild` builds

> globalThis/Symbol.for('esbuild')

I like the thought of using a shared state module, which is what this makes me think of. That'd avoid the way `globalThis` voilates our unit testing & general coding principles
That said, if we needed to attach information to a plugin instantiation, do you think it'd be tentatively future-safe to do something like `plugin.setup({...build, [Symbol.for('__my-context__'): myContext ]})`?

Thanks for your input—we love `esbuild` & are grateful for all the work
